### ISSUE TITLE
Incompatible shape for array

### ISSUE THREAD

**Issue by TomMaullin** [numba/numba#5433]
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

I have written the following function below. 
```
# ============================================================================
#
# This function takes in a matrix and vectorizes it (i.e. transforms it
# to a vector of each of the columns of the matrix stacked on top of
# one another).
#
# ============================================================================
@jit(nopython=True)
def mat2vec2D(matrix):
  
  #Return vectorised matrix
  return(matrix.transpose().reshape(matrix.shape[0]*matrix.shape[1],1))
```

When I run it without `numba` it runs well and correctly. When I run it with `numba` it gives me this error:

```
NotImplementedError: incompatible shape for array
```

**Comment by stuartarchibald** [comment]
Thanks for the report. Please could you provide more information about what caused this problem:

1. The code that you were running (as a minimal working reproducer, if you are unsure how to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).
2. Any data needed to run the code.
3. The full trace back of the exception.

Once we have a reproducer someone can take a look and work on a fix. Thanks for your help.

**Comment by TomMaullin** [comment]
Ah, of course, sure!

```
# Example matrix
matrix = np.arange(9).reshape([3,3])

# Expected answer
expected_output = np.array([0,3,6,1,4,7,2,5,8]).reshape([9,1])

# Test whether the function gave the correct result.
testVal = np.allclose(mat2vec2D(matrix), expected_output)
```

**Comment by esc** [comment]
@TomMaullin thanks, the self contained reproducer follows:

```
from numba import jit
import numpy as np

@jit(nopython=True)
def mat2vec2D(matrix):
  return(matrix.transpose().reshape(matrix.shape[0]*matrix.shape[1],1))

# Example matrix
matrix = np.arange(9).reshape([3,3])

# Expected answer
expected_output = np.array([0,3,6,1,4,7,2,5,8]).reshape([9,1])

# Test whether the function gave the correct result.
testVal = np.allclose(mat2vec2D(matrix), expected_output)
```

**Comment by esc** [comment]
And here is the traceback:

```
Traceback (most recent call last):
  File "foo025.py", line 15, in <module>
    testVal = np.allclose(mat2vec2D(matrix), expected_output)
NotImplementedError: incompatible shape for array
```

**Comment by esc** [comment]
@TomMaullin it will probably make sense to check what the shape should be and what it is instead.

**Comment by stuartarchibald** [comment]
Suspect it's a bug to do with Fortran ordering arising from the transpose.

**Comment by h3z** [comment]
same problem

**Comment by andrewrgarcia** [comment]
@TomMaullin I had a similar problem. 
But I believe I just solved it. See my solution below 

`from numba import jit
import numpy as np

@jit(nopython=True)
def mat2vec2D(matrix):
    y,x = matrix.shape
    return np.transpose(matrix).copy().reshape(y*x,1)

# Example matrix
matrix = np.arange(9.).reshape((3,3))
# Expected answer
expected_output = np.array([0,3,6,1,4,7,2,5,8]).reshape([9,1])

# Test whether the function gave the correct result.
testVal = np.allclose(mat2vec2D(matrix), expected_output)
print(testVal)`
Output: True 

Regards,
Andrew

**Comment by andrewrgarcia** [comment]
@TomMaullin I had a similar problem, but I believed I just solved it. See my solution below.
~ Andrew
```
from numba import jit
import numpy as np

@jit(nopython=True)
def mat2vec2D(matrix):
    y,x = matrix.shape
    return np.transpose(matrix).copy().reshape(y*x,1)

# Example matrix
matrix = np.arange(9.).reshape((3,3))
# Expected answer
expected_output = np.array([0,3,6,1,4,7,2,5,8]).reshape([9,1])

# Test whether the function gave the correct result.
testVal = np.allclose(mat2vec2D(matrix), expected_output)
print(testVal)
```
output: True

**Comment by DFEvans** [comment]
The issue still seems to exist - I ran into this myself today, and it turned out that providing a Fortran-aligned array was also the cause for me.

I think @andrewrgarcia's workaround is just to do an explicit `copy` inside the Numba code (presumably forcing it to be C-aligned?), while mine was to ensure my input as C-aligned before calling Numpy, but there's still a bug (or, at least, a need for a better error message/documentation explaining that Numba dislikes Fortran-aligned arrays).

**Comment by andrewrgarcia** [comment]
> The issue still seems to exist - I ran into this myself today, and it turned out that a Fortran-aligned array was also the cause for me.
> 
> 
> 
> I think @andrewrgarcia's workaround is just to do an explicit `copy` inside the Numba code (presumably forcing it to be C-aligned?), while mine was to ensure my input as C-aligned before calling Numba, but there's still a bug (or, at least, a need for a better error message/documentation explaining that Numba dislikes Fortran-aligned arrays).

Interesting. Perhaps you can offer us with a better solution?

**Comment by DFEvans** [comment]
I assume the original reproduction cases above still "work", so it's a case of tracking down the bug. Of course, I don't expect that to change quickly (it needs the time of one of the few contributors with enough knowledge to fix it!), but the earlier comments in May seemed to suggest that this issue should be closed because a user-side workaround exists, which doesn't seem to be how the Numba community usually deals with bugs.

**Comment by TomMaullin** [comment]
I have just realised I never replied to this. Apologies, I must admit I am still unsure what caused this issue and how it was resolved.

**Comment by duburcqa** [comment]
I just ran into the same issue, for the same reason. I would suggest rewording the error message to state that the array cannot be flattened by reference due because of fortran ordering.
